Dynamic voting (needs testing)

Putnam3145 opened the pull request:
## About The Pull Request

Adds an option to allow voting on threat level curve of the Dynamic game mode (or just force extended as usual), rather than simply voting between extended/secret. Keeps extended/secret vote as an option.

With dynamic voting enabled, it gives the following options:

Extended
Very calm
Somewhat calm
Balanced
Somewhat chaotic
Very chaotic

These votes determine the Dynamic game mode's threat generation curve; the more chaotic the votes, the higher the average threat level, and the more disagreement there is on how threatening it should be, the more variance there will be in the possible threat levels. Moreover, if the majority of people vote extended the mode will force extended. If the majority do *not* vote for extended, then all votes for extended are counted as "Very calm" votes.

## Why It's Good For The Game

Dynamic is good actually.

## Changelog
:cl:
add: Dynamic voting
config: Added DYNAMIC_VOTING to game_options
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Putnam3145 committed:
Dynamic voting (draft)

Putnam3145 committed:
grammar

Putnam3145 committed:
make extended vote majority instead of plurality

Putnam3145 committed:
rename indifferent to balanced

also, remember ur spaces

Ghommie commented on code/controllers/subsystem/vote.dm:
```diff
@@ -146,6 +152,17 @@ SUBSYSTEM_DEF(vote)
+				if(GLOB.dynamic_forced_extended)
+					return message_admins("Dynamic extended has been voted for.")
+				var/mean = (choices["extended"]*-2+choices[PEACE2]*-2+choices[PEACE1]*-1+choices[CHAOS1]*1+choices[CHAOS2]*2)/voted.len
+				var/variance=(((-2-mean)**2)*choices["extended"]+((-2-mean)**2)*choices[PEACE2]+((-1-mean)**2)*choices[PEACE1]+((0-mean)**2)*choices[BALANCED]+((1-mean)**2)*choices[CHAOS1]+((2-mean)**2)*choices[CHAOS2])/voted.len //Sorry. Im sorry. Im trying to remove it
```
Keep in mind this will give off a division by zero error in the unlikely occurrence of no one voting.

Putnam3145 replied:
https://github.com/Citadel-Station-13/Citadel-Station-13/pull/9557/commits/0a000f3165875ace0d8b89ef909cadb59170a5a2

Ghommie commented on code/controllers/subsystem/vote.dm:
```diff
@@ -213,6 +230,8 @@ SUBSYSTEM_DEF(vote)
 			if("roundtype") //CIT CHANGE - adds the roundstart secret/extended vote
 				choices.Add("secret", "extended")
+			if("dynamic")
+				choices.Add("extended",PEACE2,PEACE1,BALANCED,CHAOS1,CHAOS2)
```
You could as well add a "secret" option here too, but I doubt both non-dynamics are likely to be ever choosen since they still require over fifty percent of all votes to trigger.

Putnam3145 replied:
Yeah, that's a worry--could make it so that a sufficiently low mean makes it do dynamic_classic_secret.

Putnam3145 replied:
Upon really checking the code in the process of #9582, looks like secret's basically irrelevant as long as threat level's low, and I can't really think of a way to take those votes into account for if dynamic ends up winning anyway (-1? -2 like extended?), so I'll hold off on this until testing.

Ghommie commented on code/controllers/configuration/entries/game_options.dm:
```diff
@@ -364,3 +364,5 @@
 	config_entry_value = 72000
 	min_val = 0
+
+/datum/config_entry/flag/dynamic_voting
```
An admin "verb" should be added to switch on or off dynamic voting for a given round through CONFIG_SET (modified configs do not carry over successive rounds iirc). It'd be sad if this ended up backend only.

Putnam3145 replied:
I've been an admin but I'm not sure where that would go. Game panel? Secrets?

Ghommie replied:
I think somewhere easily accessible like "server" or "fun" tabs.
Since it currently would have to be hastily toggled before gamemode voting happens to work, it shouldn't be added to the secrets or game panels.
Though I suggest you ask @kevinz000, who is an administator, unlike me.

Putnam3145 replied:
Right, "server" tab makes most sense, I'll do it

Putnam3145 replied:
This sufficient? https://github.com/Citadel-Station-13/Citadel-Station-13/pull/9557/commits/da7d4345b5bdcaa07662de1ef5d9abe15329cb23

Putnam3145 replied:
Upon testing: it's sufficient.

Putnam3145 committed:
Division by zero.

Putnam3145 committed:
Added a toggle dynamic vote verb for admins.

Putnam3145 committed:
concept: actually let admins do it

Ghommie commented on code/modules/admin/admin.dm:
```diff
@@ -686,6 +686,19 @@
+	set category = "Server"
+	set desc="Switches between secret/extended and dynamic voting"
+	set name="Toggle Dynamic Vote"
+	CONFIG_SET(flag/dynamic_voting,!CONFIG_GET(flag/dynamic_voting))
```
Just state a variable `dynamic_mode` (its name can be however you want) with value `!CONFIG_GET(flag/dynamic_voting)` and replace all the following `CONFIG_GET` with it. More efficient.

Putnam3145 replied:
https://github.com/Citadel-Station-13/Citadel-Station-13/pull/9557/commits/296ce4fb78b45fbc6f088591420d5967068b26b2

Ghommie commented on code/modules/admin/admin.dm:
```diff
@@ -686,6 +686,19 @@
+		to_chat(world, "<B>Vote is now between extended and dynamic chaos.</B>")
+	else
+		to_chat(world, "<B>Vote is now between extended and secret.</B>")
+	log_admin("[key_name(usr)] toggled dynamic voting.")
```
Replace this line with the following:
```
	log_admin("[key_name(usr)] [dynamic_mode ? "enabled" : "disabled"] dynamic voting.")
```

Putnam3145 replied:
https://github.com/Citadel-Station-13/Citadel-Station-13/pull/9557/commits/296ce4fb78b45fbc6f088591420d5967068b26b2

Putnam3145 committed:
suggested changes

Putnam3145 committed:
Added dynamic vote logging

Putnam3145 committed:
s

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into super-special-awesome-dynamic

Putnam3145 committed:
Makes malf AI its own ruleset in dynamic.

Putnam3145 committed:
Makes sure that actually works.

Putnam3145 committed:
Reduced vote to just extended vs calm vs chaos.

Putnam3145 committed:
Also deobfuscates dynamic vote

Putnam3145 committed:
Okay, just peace vs chaos.

Putnam3145 committed:
actually it's gotta be delayed

Putnam3145 committed:
Makes summon (X) cost threat in dynamic

Putnam3145 committed:
WOW THIS IS IMPORTANT

Putnam3145 committed:
Actually added stuff to the config.

Putnam3145 committed:
hmm, also important

Putnam3145 committed:
disable ling whoops.

Putnam3145 committed:
balance changes (oh no)

Putnam3145 committed:
makes all enemy-required rulesets require more enemies

Putnam3145 committed:
ton of unsolicited balance changes

1. DYNAMIC_POP_PER_REQUIREMENT increased from 6 to 9, to account for ERPers by the playercount stuff
2. increased cost of cult, nuclear, revs a bit
3. Made extended slightly more likely at all pop levels, ramping up at ever lower (doesn't matter nearly as much below 25)
4. allowed clock cult

Putnam3145 committed:
Made the curve make more sense with dichotomic voting.

Putnam3145 committed:
Merge branch 'malf-dynamic' into super-special-awesome-dynamic

Putnam3145 committed:
actually fuck this config's its own file now

Putnam3145 committed:
Merge branch 'wizard-dynamic' into super-special-awesome-dynamic

Putnam3145 committed:
and finished the wizard change merge

Putnam3145 committed:
removed blood brothers from dynamic

Putnam3145 committed:
FULLY removed BB from dynamic

Putnam3145 committed:
this is probably required for compile

Putnam3145 committed:
this too, wowee

Putnam3145 committed:
Made hijack require threat threshold, spend threat

Putnam3145 committed:
fewer double roundstart antags

Putnam3145 committed:
made glorious death spend threat

GrayRachnid commented on code/controllers/subsystem/vote.dm:
```diff
@@ -379,4 +396,10 @@ SUBSYSTEM_DEF(vote)
+#undef PEACE1
+#undef BALANCED
+#undef CHAOS1
+#undef CHAOS2
```
Some of these don't exist anymore

Putnam3145 replied:
fixed in https://github.com/Citadel-Station-13/Citadel-Station-13/pull/9557/commits/85fd53d61595a0bc5ce0dce467b89278b8678341

GrayRachnid commented on config/game_options.txt:
```diff
@@ -699,3 +557,6 @@ MONKEYCAP 64
 ## Uncomment to use TG-style combat
 #DISABLE_STAMBUFFER
+
+#Replaces standard extended/secret dichotomy with extended and 5 levels of chaos using dynamic.
```
comment needs to change

Putnam3145 replied:
fuixed in https://github.com/Citadel-Station-13/Citadel-Station-13/pull/9557/commits/3a48b7238090181a90e93a6affbe25efd9423f32

Putnam3145 committed:
wow, .len, very cool

Putnam3145 committed:
no more supercults/supernukies in dynamic

Putnam3145 committed:
actually let's add the extended voting back in, see how that does

Putnam3145 committed:
let's try making the votes SWING

Putnam3145 committed:
fixed a comment

Putnam3145 committed:
tweaked the numbers and choices a little.

Putnam3145 committed:
Hacked together a default for second_rule_req and third_rule_req

Putnam3145 committed:
4-'ing twice

Putnam3145 committed:
changed misleading message

Putnam3145 committed:
massive rebalance

Putnam3145 committed:
make it calm vs chaotic again.

Putnam3145 committed:
more tweaks to requirements

Putnam3145 committed:
Added logging to wizard/traitor extra threat spending.

Putnam3145 committed:
wrong paren placement wowee

Putnam3145 committed:
Hardcoded default config values.

Putnam3145 committed:
Made curve width formula be less stupid, less admin spam

Putnam3145 committed:
made warops part of the dynamic system, in dynamic

Putnam3145 committed:
Added a preference that auto-votes at extremely low weight.

Putnam3145 committed:
Updated flavor text to better represent balance.

Putnam3145 committed:
Made antag injection happen more often, initial injection later.

Putnam3145 committed:
ported (commented out) dynamic vampire ruleset from 413

Putnam3145 committed:
started transferring some events from events to dynamic antags

Putnam3145 committed:
Dynamic controls lots of events now (when active)!

Putnam3145 committed:
wowee

Putnam3145 committed:
errors hooray

Putnam3145 committed:
important organization comments

Putnam3145 committed:
sorry abductors ilu

Putnam3145 committed:
New stuff to config

Putnam3145 committed:
more repeatability!

Putnam3145 committed:
massively boost midround injection amounts

Putnam3145 committed:
Made roundstart traitors always possible at >=60 threat

Putnam3145 committed:
might be better if events actually run

Putnam3145 committed:
Made the roundstart flavor make a bit more sense.

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into super-special-awesome-dynamic

Putnam3145 committed:
okay none of that i guess

Putnam3145 committed:
Made vote consensus way stronger

Putnam3145 committed:
Improved roundstart report flavor

Putnam3145 committed:
MORE! MORE! MORE!

Putnam3145 committed:
wow, totally the wrong proc, very cool

Putnam3145 committed:
Better roundend logging

Putnam3145 committed:
Nuke disk being still now generates threat

Putnam3145 committed:
less terrible admin spam

Putnam3145 committed:
m o r e

Putnam3145 committed:
also pirates

Putnam3145 committed:
aaa config

Putnam3145 committed:
Assassination is less likely at lower threats.

Putnam3145 committed:
mm, yes, compile BEFORE committing

Putnam3145 committed:
i overtuned it. time to go back down

Putnam3145 committed:
lowered maximum curve width a good deal

Putnam3145 committed:
reobfuscates dynamic vote

Putnam3145 committed:
Weighted antags higher, made zero-antag midround injection chance higher

Putnam3145 committed:
Better logging of why-no-roundstart

Putnam3145 committed:
forces antag if all antags are dead

Ghommie committed:
Initial stuff.

Ghommie committed:
fixeronis.

Ghommie committed:
That's it: #46027, (a little bit of) #46132, #47100, #47341, #47351, #47395, #47457, #47491.

Putnam3145 committed:
Merged upstream dynamic fixes ported by Ghommie.

Putnam3145 committed:
all roundstarts are delayed now

Putnam3145 committed:
um.

Putnam3145 committed:
Fixing some logging

Putnam3145 committed:
Further logging fixes

Putnam3145 committed:
Editing stuff to let it compile (see description)

Malf AI is no longer available as a roundstart ruleset because antag assignment happens before job assignment. On secret, this is not a problem, but on dynamic, there is no way to make malfunctioning AI limited to certain threat levels except by making it a delayed ruleset. Unfortunately, with the upstream changes, delayed rulesets were removed--all roundstart rulesets are delayed, and if it were to roll malf AI, it would simply error, make a runtime stack trace and, in general, shit the bed. There would be no roundstart antag that round, which is *kind of* a problem I'm desperate to fix. So, no more roundstart malf--it's either that or some rounds having no roundstart antags at all.

Putnam3145 committed:
Removed a comment I only kept around for copy+pasting.

Putnam3145 committed:
balancing (hopefully more roundstart antags)

Putnam3145 committed:
ok from_ghosts

the worst part is that putting this boilerplate as a base part of the proc probably isn't even advisable. I hate it.

Putnam3145 committed:
elligible more like illegible

Putnam3145 committed:
added weights, repeatable weight decreases

Putnam3145 committed:
made forced midround antags increase weight by their cost

This is to make sure that threat is really, really used throughout a round, or at least to encourage it to be--people are rightfully annoyed at 70 threat level rounds just not using the vast majority of it.

Putnam3145 committed:
Let's make those roundstarts more likely.

Putnam3145 committed:
the "latejoiners count too" commit

Putnam3145 committed:
Rethink events (see desc)

Made events their own ruletype, made the duplicated-with-different-threats events are now one event with shared values that can scale theirselves up.

Also fixed **MIDROUND ANTAGS NOT WORKING AT ALL BECAUSE THEY WERE LOOKING FOR ENEMIES ONLY AMONG PLAYERS ELIGIBLE FOR THE ANTAG WTF**

Putnam3145 committed:
updated an error to actually be right

Putnam3145 committed:
more logging.

GrayRachnid commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -465,3 +482,244 @@
+	required_candidates = 1
+	weight = 4
+	cost = 15
+	requirements = list(101,101,101,90,80,70,60,50,40,30)
```
why is ninja so expensive to run at midpop

Putnam3145 replied:
didn't get caught in my balance passes, probably, will tweak

Putnam3145 committed:
lowered admin spam

Putnam3145 committed:
whoops forgot initialize cooldown

Putnam3145 committed:
Increased event cooldown

Putnam3145 committed:
this is why static typing is good and should be used more

GrayRachnid commented on config/dynamic_config.txt:
```diff
@@ -0,0 +1,299 @@
+
+## Latejoin antags
+DYNAMIC_HIGH_POPULATION_REQUIREMENT LATEJOIN_TRAITOR 15
+DYNAMIC_HIGH_POPULATION_REQUIREMENT LATEJOIN_REVOLUTION 50
```
forgot to drop this down to 20

GrayRachnid commented on config/dynamic_config.txt:
```diff
@@ -0,0 +1,299 @@
+
+## Midround antags
+DYNAMIC_HIGH_POPULATION_REQUIREMENT MIDROUND_TRAITOR 15
+DYNAMIC_HIGH_POPULATION_REQUIREMENT MIDROUND_MALF_AI 35
```
this similarly should be at 30 compared to above

GrayRachnid commented on config/dynamic_config.txt:
```diff
@@ -0,0 +1,299 @@
+## Midround antags
+DYNAMIC_HIGH_POPULATION_REQUIREMENT MIDROUND_TRAITOR 15
+DYNAMIC_HIGH_POPULATION_REQUIREMENT MIDROUND_MALF_AI 35
+DYNAMIC_HIGH_POPULATION_REQUIREMENT MIDROUND_WIZARD 50
```
this too

Putnam3145 committed:
blacklisted blob from clock cult, latejoin revs now higher req threat

Putnam3145 committed:
holy shit

Putnam3145 committed:
Made results of vote semi-public

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into super-special-awesome-dynamic

Putnam3145 committed:
why is this HERE

Putnam3145 committed:
Made roundstart

Putnam3145 committed:
this might help with the disk weirdness

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into super-special-awesome-dynamic

Putnam3145 committed:
i'm like 90% sure this is what it's supposed to be.

Putnam3145 committed:
Removed janky nuke disk threat increase

Putnam3145 committed:
Made s?laughter demon and apprentice cost threat, too.

Putnam3145 committed:
Reduced RNG a bit more

Putnam3145 committed:
Added some more blocking rules (whoops)

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into super-special-awesome-dynamic

Putnam3145 committed:
Made staff of change require/cost threat
